src/mlpack/methods/matrix_completion/matrix_completion.hpp

```diff
@@ -0,0 +1,150 @@
+ * following optimization problem (semi-definite program) is solved to fill in
+ * the remaining unknown values of X
+ *
+ *   min ||X||_* subj to X_ij = M_ij
+ *
+ * where ||X||_* denotes the nuclear norm (sum of singular values of X).
+ *
+ * For a theoretical treatment of the conditions necessary for exact recovery,
+ * see the following paper:
+ *
+ *   A Simpler Appoarch to Matrix Completion.
+ *   Benjamin Recht. JMLR 11.
+ *   http://arxiv.org/pdf/0910.0651v2.pdf
+ *
+ * An example of how to use this class is shown below:
+ *
+ * @code
+ * size_t m, n;        // size of unknown matrix
+ * arma::umat indices; // contains the known indices [2 x n_entries]
+ * arma::vec values;   // contains the known values [n_entries]
+ *
+ * MatrixCompletion mc(m, n, indices, values);
+ * mc.Recover();
+ * mc.Recovered();     // access completed matrix
+ * @endcode
+ *
+ * @see LRSDP
+ */
+class MatrixCompletion
+{
+ public:
+
+  /**
+   * Construct a matrix completion problem, specifying the maximum rank of the
+   * solution.
+   *
+   * @param m Number of rows of original matrix.
+   * @param n Number of columns of original matrix.
+   * @param indices Matrix containing the indices of the known entries (must be
+   *    [2 x p]).
+   * @param values Vector containing the values of the known entries (must be
+   *    length p).
+   * @param r Maximum rank of solution.
+   */
+  MatrixCompletion(const size_t m,
+                   const size_t n,
+                   const arma::umat& indices,
+                   const arma::vec& values,
+                   const size_t r);
+
+  /**
+   * Construct a matrix completion problem, specifying the initial point of the
+   * optimization.
+   *
+   * @param m Number of rows of original matrix.
+   * @param n Number of columns of original matrix.
+   * @param indices Matrix containing the indices of the known entries (must be
+   *    [2 x p]).
+   * @param values Vector containing the values of the known entries (must be
+   *    length p).
+   * @param initialPoint Starting point for the SDP optimization.
+   */
+  MatrixCompletion(const size_t m,
+                   const size_t n,
+                   const arma::umat& indices,
+                   const arma::vec& values,
+                   const arma::mat& initialPoint);
+
+  /**
+   * Construct a matrix completion problem.
+   *
+   * @param m Number of rows of original matrix.
+   * @param n Number of columns of original matrix.
+   * @param indices Matrix containing the indices of the known entries (must be
+   *    [2 x p]).
+   * @param values Vector containing the values of the known entries (must be
+   *    length p).
+   */
+  MatrixCompletion(const size_t m,
+                   const size_t n,
+                   const arma::umat& indices,
+                   const arma::vec& values);
+
+  /**
+   * Solve the underlying SDP to fill in the remaining values.
+   */
+  void Recover();
```

Comment:
Personall I think `void Recover(arma::mat& output)` is a better option here, and maybe we ought to call it `Apply` and attempt to make it similar to the interface required by CF's FactorizerType class: http://mlpack.org/doxygen.php?doc=classmlpack_1_1cf_1_1CF.html#_details .  That may take a little more thinking, though, and would probably require refactoring the class so the LRSDP constraints are set when `Recover()` is called.